client/worker.go

```diff
@@ -0,0 +1,241 @@
+	"go.opencensus.io/trace"
+
+	sapmpb "github.com/signalfx/sapm-proto/gen"
+)
+
+const (
+	defaultRateLimitingBackoffSeconds = 8
+	headerAuthToken                   = "X-SF-Token"
+	headerRetryAfter                  = "Retry-After"
+	headerContentEncoding             = "Content-Encoding"
+	headerContentType                 = "Content-Type"
+	headerValueGZIP                   = "gzip"
+	headerValueXProtobuf              = "application/x-protobuf"
+)
+
+type worker struct {
+	authToken  string
+	client     *http.Client
+	endpoint   string
+	gzipWriter *gzip.Writer
+	maxRetries uint64
+	doneCh     chan struct{}
+}
+
+func newWorker(client *http.Client, endpoint string, authToken string, maxRetries uint64) *worker {
+	w := &worker{
+		authToken:  authToken,
+		client:     client,
+		endpoint:   endpoint,
+		gzipWriter: gzip.NewWriter(nil),
+		maxRetries: maxRetries,
+		doneCh:     make(chan struct{}),
+	}
+	return w
+}
+
+func (w *worker) export(ctx context.Context, batch *jaegerpb.Batch) error {
+	_, span := trace.StartSpan(ctx, "export")
+	defer span.End()
+
+	if len(batch.Spans) == 0 {
+		span.AddAttributes(trace.BoolAttribute("empty_batch", true))
+		return nil
+	}
+
+	sr, err := w.prepare(ctx, batch)
+	if err != nil {
+		recordEncodingFailure(ctx, sr)
+		span.SetStatus(trace.Status{
+			Code: trace.StatusCodeInvalidArgument,
+		})
+		return err
+	}
+
+	retries := uint64(0)
+
+	for {
+		err := w.send(ctx, sr)
+		if err == nil {
+			recordSuccess(ctx, sr)
+			break
+		}
+		recordSendFailure(ctx, sr)
+
+		if err.Permanent {
+			span.SetStatus(trace.Status{
+				Code: ocTrancelator.OCStatusCodeFromHTTP(int32(err.StatusCode)),
+			})
+			recordDrops(ctx, sr)
+			return err
+		}
+
+		if err.RetryDelaySeconds > 0 {
+			if retries >= w.maxRetries {
+				span.SetStatus(trace.Status{
+					Code: trace.StatusCodeDeadlineExceeded,
+				})
+				recordDrops(ctx, sr)
+				err.Permanent = true
+				return err
+			}
+			time.Sleep(time.Duration(err.RetryDelaySeconds) * time.Second)
+			retries++
+			continue
+		}
+		return err
+	}
+	return nil
+}
+
+func (w *worker) send(ctx context.Context, r *sendRequest) *ErrHTTPSend {
+	_, span := trace.StartSpan(ctx, "export")
+	defer span.End()
+
+	req, err := http.NewRequest("POST", w.endpoint, bytes.NewBuffer(r.message))
+	if err != nil {
+		span.SetStatus(trace.Status{
+			Code:    trace.StatusCodeInvalidArgument,
+			Message: err.Error(),
+		})
+		return &ErrHTTPSend{Err: err, Permanent: true}
+	}
+	req.Header.Add(headerContentType, headerValueXProtobuf)
+	req.Header.Add(headerContentEncoding, headerValueGZIP)
+	if w.authToken != "" {
+		req.Header.Add(headerAuthToken, w.authToken)
+	}
+
+	resp, err := w.client.Do(req)
+	if err != nil {
+		span.SetStatus(trace.Status{
+			Code:    trace.StatusCodeInternal,
+			Message: err.Error(),
+		})
+		return &ErrHTTPSend{Err: err}
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode >= 200 && resp.StatusCode <= 299 {
+		return nil
```

Comment:
If we ever want the underlying tcp connection to be reused by the http.Client, then response body from http.Client.Do() must be read to EOF and closed.  Not sure if this is a concern for this client or not.  I've used `io.Copy(ioutil.Discard, resp.Body)` before in another client to ensure the body is read.

Reply:
I thought this was only true for Go <1.8? Is it still the case for >=1.12 ?

Reply:
OK it's still unclear. I added a line to read upto ~256kb.

Reply:
I think it's still relevant.  The doc for client.Do says it's still needed.  Then again, maybe the doc is out of date?

Reply:
I added in any case but with a limit on how much it reads which should take care of our use case.